### PR TITLE
feat: support stateful precompiled contracts

### PR DESCRIPTION
**Pull request by mmsqe**
Updates: #1116

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Comment by 0xadu** [comment]
great work! Is this blocked by some PRs in the upstream repo? Anything that I could help to make it usable?

**Comment by mmsqe** [comment]
> great work! Is this blocked by some PRs in the upstream repo? Anything that I could help to make it usable?

Mainly waiting for sdk [refactor](https://github.com/cosmos/cosmos-sdk/pull/14444)

File: x/evm/statedb/statedb.go [hunk 1]
```diff
@@ -80,11 +86,24 @@ func New(ctx sdk.Context, keeper Keeper, txConfig TxConfig) *StateDB {
...
+}
+
 // Keeper returns the underlying `Keeper`
 func (s *StateDB) Keeper() Keeper {
 	return s.keeper
 }
 
+// AppendJournalEntry allow external module to append journal entry,
+// to support snapshot revert for external states.
+func (s *StateDB) AppendJournalEntry(entry JournalEntry) {
```
Comment (yihuang) [inline x/evm/statedb/statedb.go:103]: Is this method used anywhere?

Comment (mmsqe) [reply x/evm/statedb/statedb.go:103]: No, can clean up some interface now

**Pull request closed without merging**